src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -304,8 +330,69 @@ size_t runtime::gpu::CUDNNEmitter::build_reduce_forward(const cudnnReduceTensorO
                                                           beta,
                                                           output_desc,
                                                           reduce_buffer));
+                        gpu::invoke_primitive(m_ctx, convert_idx, &workspace_indices_ptr, outputs);
                         debug_sync();
                     }});
+                }
+            }
+            else
+            {
+                if (input_type == element::i32)
+                {
+                    size_t input_idx = allocator.reserve_workspace(shape_size(input_shape) *
+                                                                   modified_input_type.size());
+                    auto& cuda_emitter = m_primitive_emitter->get_cuda_emitter();
+                    auto convert_input_idx = cuda_emitter->build_elementwise<op::Convert>(
+                        {input_type.c_type_string(), modified_input_type.c_type_string()},
+                        input_shape);
+
+                    reduce.reset(new gpu::primitive{
+                        [=, &desc, &input_desc, &output_desc](void** inputs, void** outputs) {
+                            void* input_f64_ptr =
+                                runtime::gpu::invoke_memory_primitive(m_ctx, input_idx);
+                            void* workspace_ptr =
+                                runtime::gpu::invoke_memory_primitive(m_ctx, workspace_idx);
+                            void* reduce_buffer =
+                                runtime::gpu::invoke_memory_primitive(m_ctx, reduce_buffer_idx);
+                            gpu::invoke_primitive(m_ctx, convert_input_idx, inputs, &input_f64_ptr);
+                            CUDNN_SAFE_CALL(cudnnReduceTensor(*m_ctx->cudnn_handle,
+                                                              desc,
+                                                              outputs[0],
+                                                              indices_size,
+                                                              workspace_ptr,
+                                                              workspace_size,
+                                                              alpha,
+                                                              input_desc,
+                                                              input_f64_ptr,
+                                                              beta,
+                                                              output_desc,
+                                                              reduce_buffer));
+                            debug_sync();
+                        }});
+                }
+                else
+                {
+                    reduce.reset(new gpu::primitive{
```

Comment:
nit: maybe use callbacks to simplify the logic: https://github.com/NervanaSystems/ngraph/compare/amoitra/arg_reduce_int32...cs/amoitra/arg_reduce_int32